lib/jobs.ts

```diff
@@ -0,0 +1,217 @@
+    return bindings.map((binding) => binding['reportId'].value);
+  } else {
+    return [];
+  }
+}
+
+async function getNextScheduledJob() {
+  const result = await query(`
+  PREFIX mu: <http://mu.semte.ch/vocabularies/core/>
+  PREFIX ext: <http://mu.semte.ch/vocabularies/ext/>
+  PREFIX dct: <http://purl.org/dc/terms/>
+  PREFIX adms: <http://www.w3.org/ns/adms#>
+
+  SELECT ?uri ?id ?status
+  WHERE {
+    GRAPH <${config.job.graph}> {
+      VALUES ?status {
+        ${sparqlEscapeUri(config.job.statuses.scheduled)}
+      }
+      ?uri a ext:ReportGenerationJob ;
+           mu:uuid ?id ;
+           dct:created ?created ;
+           adms:status ?status .
+      FILTER NOT EXISTS {
+        ?job a ext:ReportGenerationJob ;
+           adms:status ${sparqlEscapeUri(config.job.statuses.ongoing)} .
+      }
+    }
+  } ORDER BY ASC(?created) LIMIT 1`);
+
+  const bindings = result.results.bindings;
+  if (bindings.length === 1) {
+    let job = {
+      id: bindings[0]['id'].value,
+      uri: bindings[0]['uri'].value,
+      status: bindings[0]['status'].value,
+    };
+    job.reportIds = await getReportIds(job);
+    return job;
+  } else {
+    return null;
+  }
+}
+
+export async function getJob(jobId) {
+  const result = await query(`
+  PREFIX mu: <http://mu.semte.ch/vocabularies/core/>
+  PREFIX ext: <http://mu.semte.ch/vocabularies/ext/>
+  PREFIX dct: <http://purl.org/dc/terms/>
+  PREFIX adms: <http://www.w3.org/ns/adms#>
+
+  SELECT ?uri ?id ?status ?created ?modified
+  WHERE {
+    GRAPH <${config.job.graph}> {
+      ?uri a ext:ReportGenerationJob ;
+           mu:uuid ${sparqlEscapeString(jobId)} ;
+           dct:created ?created ;
+           dct:modified ?modified ;
+           adms:status ?status .
+    }
+  } ORDER BY ASC(?created) LIMIT 1`);
+
+  const bindings = result.results.bindings;
+  if (bindings.length === 1) {
+    let job = {
+      id: jobId,
+      uri: bindings[0]['uri']?.value,
+      status: bindings[0]['status']?.value,
+      created: bindings[0]['created']?.value,
+      modified: bindings[0]['modified']?.value,
+    };
+    job.reportIds = await getReportIds(job);
+    return job;
+  } else {
+    return null;
+  }
+}
+
+async function updateJobStatus(uri, status) {
+  await update(`
+  PREFIX dct: <http://purl.org/dc/terms/>
+  PREFIX adms: <http://www.w3.org/ns/adms#>
+
+  DELETE WHERE {
+    GRAPH <${config.job.graph}> {
+        ${sparqlEscapeUri(uri)} dct:modified ?modified ;
+             adms:status ?status.
+    }
+  }
+
+  ;
+
+  INSERT DATA {
+    GRAPH <${config.job.graph}> {
+        ${sparqlEscapeUri(uri)} dct:modified ${sparqlEscapeDateTime(new Date())};
+             adms:status ${sparqlEscapeUri(status)}.
+    }
+  }`);
+}
+
+async function executeJob(job) {
+  try {
+    await updateJobStatus(job.uri, config.job.statuses.ongoing);
+    for (const reportId of job.reportIds) {
+      const fileMeta = await generateReport(reportId, jobRequestHeaders[job.id]);
```

Comment:
**issue(non-blocking)**: This will never work for jobs triggered by the cron job

The queries inside `generateReport` are normal queries that expect to be triggered by a user. If the cron job triggers them, they will silently return no data, which the function doesn't cope with well.

I think we don't notice it while testing locally because creating a job immediately executes it and presumably we're not creating enough jobs at once to have them get picked up by the cron job.

When it fails, the job correctly gets the failure status, so we might be okay with this for now (we just need to check up on it manually).

Reply:
You're absolutely right. As discussed via chat, I'll look for an alternative for this, because now it's just a whole bunch of useless code.